File: solved/14/solution.c
/* projecteuler.net
 * Problem 14 
 * Q. The following iterative sequence is defined for the set of positive integers:
 *
 * n  n/2 (n is even)
 * n  3n + 1 (n is odd)
 *
 * Using the rule above and starting with 13, we generate the following sequence:
 *
 * 13  40  20  10  5  16  8  4  2  1
 * It can be seen that this sequence (starting at 13 and finishing at 1) contains 10 terms. Although it has not been proved yet (Collatz Problem), it is thought that all starting numbers finish at 1.
 *
 * Which starting number, under one million, produces the longest chain?
 */

#include <stdio.h>

/* Takes a starting point and returns the chain length */
int algo(int start){
    int new;
    int count = 1;

    while(new != 1){
        if((start % 2)) 
            new = 3 * start + 1;
        else
            new = start / 2;
        start = new;
        count++;
    }
    return count;
}

int main(){
    int old, new, sp, i = 0;

    for(i=1; i < 100; i++){
        new = algo(i);
        if(new > old){
            old = new;
            sp = i;
        }
    }
    printf("start: %i, chain: %i\n", sp, old);
    
    return 0;

}


        

File: solved/14/solution.rb
#!/usr/bin/env ruby
#
# projecteuler.net
# Solution to 14
#
# Start point: 837799
# Chain Length: 525
#

class Solver
    def self.algo(start)
        new = 0
        count = 1
        while new != 1 do
            if (start % 2) == 0
                new = start / 2
            else
                new = 3 * start + 1
            end
            start = new;
            count += 1
        end
        count
    end

    def self.run(range_start, range_end)
        old = 0
        new = 0
        sp = 0

        range_start.upto range_end do |i|
            new = algo(i)
            if new > old
                old = new
                sp = i
            end
        end
        puts "Start point: #{sp}\nChain Length: #{old}"
    end
end

#Solver.run(ARGV[0].to_i, ARGV[1].to_i)
Solver.run(750000, 1000000)

